File: fortran/recursivecollatz.f90
!This program will perform the Collatz conjecture
program main
      integer*16 :: maxnum
      integer*16, dimension(10) :: numlist
      integer*16, dimension(10) :: countlist
      integer*16 :: x, smallcount, counter, collatz
      integer*16 :: smallnum, indexnum, duplicate, temp
      external collatz
      maxnum = 10000
      !Zeroes out the 2 arrays
      do i = 1, 10
            numlist(i) = 0
            countlist(i) = 0
      end do

      do while(maxnum.ne.0)
            x = maxnum
            duplicate = 0
            smallnum = 1
            smallcount = countlist(1)
            indexnum = 1
            counter = collatz(x) !Calls function collatz
            do i = 1, 10
                  !Updates if current sequence greater than previous
                  if(smallcount > countlist(i)) then
                        smallcount = countlist(i)
                        smallnum = i
                  end if
                  !Check for duplicates 
                  if(countlist(i) == counter) then
                        duplicate = 1
                        indexnum = i
                  end if
            end do
            temp = smallnum
            !No duplicates and adds to the sequence
            if(counter > countlist(temp).and.duplicate == 0) then
                  numlist(temp) = maxnum
                  countlist(temp) = counter
            end if
            !If duplicate then add smaller number
            if(maxnum < countlist(indexnum).and.duplicate == 1) then
                  numlist(indexnum) = maxnum
            end if
            maxnum = maxnum - 1
      end do
      do j = 1,10
            print *, numlist(j), "  ", countlist(j)
      end do
end program main

recursive integer(kind=16) function collatz(x) result(rcounter)
            integer*16 :: x
            if(x == 1) then
                  rcounter = 0
                  return
            else if((mod(x,2)) == 1) then  !Odd number
                  rcounter = collatz((x * 3) + 1) + 1
            else  !Even number
                  rcounter = collatz(x / 2) + 1
            end if
end function collatz

File: java/recursivecollatz.java
//Program that calculates the Collatz conjecture
import java.util.*;
import java.io.*;
import java.math.BigInteger;
public class recursivecollatz
{
      public class list
      {
            long num;
            long count;
      }
      public static void main(String[] args)
      {
            long max = 10000;
            long[] numlist = new long[10];
            long[] countlist = new long[10];
            while(max != 0)
            {
                  long x = max;
                  boolean duplicate = false;
                  int smallnum = 0;
                  long smallcount = countlist[0];
                  int index = 0;
                  long count = collatz(x);  //Calls collatz method to do actual calculations
                  for(int i = 0; i < 10; i++)
                  {
                        //Updates if current sequence greater than previous sequence
                        if(smallcount > countlist[i])
                        {
                              smallcount = countlist[i];
                              smallnum = i;
                        }
                        //Check for duplicate sequence
                        if(countlist[i] == count)
                        {
                              duplicate = true;
                              index = i;
                        }
                  }
                  int temp = smallnum;
                  //No duplicates and adds in sequence if sequence greater than current 
                  if(count > countlist[temp] && duplicate == false)
                  {
                        numlist[temp] = max;
                        countlist[temp] = count;
                  }
                  //If duplicates adds the smaller number
                  if(max < countlist[index] && duplicate == true)
                        numlist[index] = max;
                  max--;
            }
            bubbleSort(numlist, countlist);
            System.out.println("Sequence sorted by integer size");
            for(int j = 0; j < 10; j++)
                  System.out.println(numlist[j] + "  " + countlist[j]);
      }
      static long collatz(long x)
      {
            //Loop to do collatz
            while(x != 1)
            {
                  if(x % 2 == 1)
                  {
                        return 1 + collatz((x * 3) + 1);
                  }
                  else
                  {
                        return 1 + collatz(x / 2);
                  }
            }
            return x - 1;
      }
      static void bubbleSort(long numlist[], long countlist[])  //Method that does bubblesort
      {
            for (int x = 0; x < 10; x++) 
                  for (int y = 0; y < 9 - x; y++) 
                        if(numlist[y] < numlist[y + 1]) 
                        {
                              long swapNum = numlist[y];
                              numlist[y] = numlist[y + 1];
                              numlist[y + 1] = swapNum;
                              long swapCount = countlist[y];
                              countlist[y] = countlist[y + 1];
                              countlist[y + 1] = swapCount;
                        }
     }  
}
